source/Conversation.cpp

```diff
@@ -442,46 +459,119 @@ const Sprite *Conversation::Scene(int node) const
+		return false;
+	const auto &data = nodes[node].elements[element];
+	if(data.conditions.IsEmpty())
+		return true;
+	return data.conditions.Test(vars);
+}
+
+
+
+// Returns true if the given node index is in the range of valid nodes for this
+// Conversation.
+bool Conversation::NodeIsValid(int node) const
+{
+	if(node < 0)
+		return false;
+	return static_cast<unsigned>(node) < nodes.size();
+}
+
+
+
+// Returns true if the given node index is in the range of valid nodes for this
+// Conversation *and* the given element index is in the range of valid elements
+// for the given node.
+bool Conversation::ElementIsValid(int node, int element) const
+{
+	if(!NodeIsValid(node))
+		return false;
+	else if(element < 0)
+		return false;
+	return static_cast<unsigned>(element) < nodes[node].elements.size();
 }
 
 
 
-// Parse the children of the given node to see if then contain any "gotos."
-// If so, link them up properly. Return true if gotos were found.
+// Parse the children of the given node to see if then contain any "gotos," or
+// "to shows." If so, link them up properly. Return true if gotos or
+// conditions were found.
 bool Conversation::LoadGotos(const DataNode &node)
 {
 	bool hasGoto = false;
+	bool hasCondition = false;
 	for(const DataNode &child : node)
 	{
-		if(hasGoto)
+		if(hasGoto && hasCondition)
 			child.PrintTrace("Warning: Ignoring extra text in conversation choice:");
 		else if(child.Size() == 2 && child.Token(0) == "goto")
 		{
 			// Each choice can only have one goto
-			Goto(child.Token(1), nodes.size() - 1, nodes.back().data.size() - 1);
-			hasGoto = true;
+			if(hasGoto)
+				child.PrintTrace("Warning: Ignoring extra endpoint in conversation choice:");
+			else
+			{
+				Goto(child.Token(1), nodes.size() - 1, nodes.back().elements.size() - 1);
+				hasGoto = true;
+			}
+		}
+		else if(child.Size() == 2 && child.Token(0) == "to" && child.Token(1) == "display")
+		{
+			// Each choice can only have one condition
+			if(hasCondition)
+				child.PrintTrace("Warning: Ignoring extra condition in conversation choice:");
+			else
+			{
+				nodes.back().elements.back().conditions.Load(child);
+				hasCondition = true;
+			}
 		}
 		else
 		{
 			// Check if this is a recognized endpoint name.
 			int index = TokenIndex(child.Token(0));
 			if(child.Size() == 1 && index < 0)
 			{
-				nodes.back().data.back().second = index;
-				hasGoto = true;
+				if(hasGoto)
+					child.PrintTrace("Warning: Ignoring extra endpoint in conversation choice:");
+				else
+				{
+					nodes.back().elements.back().next = index;
+					hasGoto = true;
+				}
 			}
 			else
-				child.PrintTrace("Warning: Expected goto or endpoint in conversation, found this:");
+				child.PrintTrace("Warning: Expected goto, to show, or endpoint in conversation, found this:");
 		}
 	}
-	return hasGoto;
+	return hasGoto || hasCondition;
+}
+
+
+
+bool Conversation::HasCondition(const DataNode &node)
```

Comment:
The function name is not great; because this function only checks if the node contains a `to display`. (But we might want to remove the function anyway...

Reply:
Maybe rename it to `HasDisplayRestriction` ?

Reply:
That name is much better. (You seem to be on a naming roll.)

Reply:
Did this for you.